### ISSUE TITLE
Query performance (table indexes are ignored)

### ISSUE THREAD

**Issue by nhenezi** [lykahb/groundhog#57]
Hello,

please take a look at `equalsOperator a b` for Postgresql: https://github.com/lykahb/groundhog/blob/6d2f68129109330e6e0d0c3d98f03e791ee7a741/groundhog-postgresql/Database/Groundhog/Postgresql.hs#L66

`a IS NOT DISTINCT FROM b` always ignores indexes, which makes queries on large table quite expensive

```
testdb=# explain ANALYZE UPDATE "public"."test_table" SET "status"='Something' WHERE "id" IS NOT DISTINCT FROM 468029;                                                                          
                                                     QUERY PLAN                                                     
--------------------------------------------------------------------------------------------------------------------
 Update on test_table  (cost=0.00..59080.88 rows=1 width=87) (actual time=544.271..544.271 rows=0 loops=1)
   ->  Seq Scan on test_table  (cost=0.00..59080.88 rows=1 width=87) (actual time=105.891..543.881 rows=1 loops=1)
         Filter: (NOT (id IS DISTINCT FROM 468029))
         Rows Removed by Filter: 2139457
 Total runtime: 544.308 ms
(5 rows)
```

Compared to a query using index:

```
testdb=# explain ANALYZE UPDATE "public"."test_table" SET "status"='Something' WHERE "id"=468029;                                                                                               
                                                             QUERY PLAN                                                              
-------------------------------------------------------------------------------------------------------------------------------------
 Update on test_table  (cost=0.43..8.45 rows=1 width=87) (actual time=0.069..0.069 rows=0 loops=1)
   ->  Index Scan using test_table_pkey on test_table  (cost=0.43..8.45 rows=1 width=87) (actual time=0.034..0.035 rows=1 loops=1)
         Index Cond: (id = 468029)
 Total runtime: 0.116 ms
(4 rows)
```

I understand that from Haskell perspective `Nothing == Nothing`, but SQL standard (and postgresql) doesn't consider any two `NULL` values equal.
cb3adcaa65b743a2dab338ffced84a2005de0213 (from 2011) introduced `IS NOT DISTINCT` as equals operator. Is there any other reason (except `NULL` behavior) that requires use `IS NOT DISTINCT` (or `<>`, `<=>` in `mysql`/`sqlite`)?

Would you be open to fixing this?
I'm new to the groundhog codebase, so I'm not sure what would be the best way to fix this. Do you have any suggestions?

In any case, thank you for this awesome library - keep up the good work!

**Comment by lykahb** [comment]
Thank you for the report. Indeed, the only reason for using IS NOT DISTINCT is NULL behavior. I did not expect these operators to affect performance so much.

It is not obvious how to fix this because at the SQL generation stage the types are not available. There are only PersistValue's. I will investigate the options to improve performance.

**Comment by MichaelXavier** [comment]
I just noticed this today and I think this can be rather serious. Is there anything in userland we can do to indicate that we really want to use an `=` operator? For lookups, it seems like the `get` function will use `=`, but if you ever need to apply updates, there is no special case for an update against a primary key, so you're left with specifying conditions. I've got a table with `~2m` row with frequent id-based updates and not using `=` is several orders of magnitude slower (up to 7s per update) and I'm fairly certain is pinning the CPU.

The only idea I can come up with ATM is to convert those write operations on that table to use `executeRaw` while this is still an issue.

**Comment by MichaelXavier** [comment]
To answer my own question, a nice workaround that doesn't require dropping to raw is to use `in_` with a singleton list. In my tests so far, postgres is smart enough to use an index in that case and produce equivalent performance for primary key lookup.

**Comment by lykahb** [comment]
@MichaelXavier Thanks for sharing the workaround. This is a bad performance issue. Now I think that Groundhog API could be closer to SQL semantics. The operators < and > have never checked for NULL so even now Groundhog does not fully mimic Haskell comparisons.

I will make ==. and /=.  rendered into regular SQL comparison operators. Their current behavior will be to the operators ===. and /== which names are similar to the identity operators in Javascript. This is a breaking change that requires at least a GHC warning pragma.

**Comment by MichaelXavier** [comment]
Hi @lykahb, I was wondering if you'd be able to follow up on this issue. I periodically get stung by this issue, and especially on large tables it makes primary key lookups and updates into sequential scans, making them many hundreds of times slower than they could be.

**Comment by lykahb** [comment]
One problem is that at the time the conditions are rendered into SQL, the types that can tell if the data is nullable, are not readily available for all kinds of expressions. It's not clear how to fix this in the general case yet.

There can be a fix for the simple case when comparing two one-column expressions of field or literal type. I'll do that soon.

**Comment by lykahb** [comment]
Fixed in 0.10
